README.md
# EmailParse

Easy parsing of emails in PHP

## Parse an email

    include_once('EmailMap.php');
    include_once('EmailParse.php');

    // $input is a raw email string
    // get $input from some other php library like a pop3 library or something

    $ep = new EmailParse();
    $email_map = $ep->parse($input);

    print_r($email_map->getTo()); // getTo() returns a list since there can be more than one to address
    echo $email_map->getFrom();

    echo $email->getSubject();
    echo $email->getPlainBody();

    print_r($email_map->getAttachments());

## Todo

Right now you have to have the full email before you can parse it, some email can be pretty big
if they have lots of attachments (or one really big attachment). So it would probably be prudent
to make `parse()` be able to take a file-pointer so it can read in the email line by line or something.
I haven't done this yet because it's never been a problem in the places I use this lib.

## License

MIT, pull requests encouraged.



CHANGELOG.md
# Changelog

3-18-07 - initial writing of the class

7-28-07 - cleaned it up and added it to the standard lib, this is now a core class

8- 15-20 -07 - made this class into a pop3 fetching class also, now it can fetch email
  from a pop3 email box, organize attachments and place them in a directory.

8-20-07 - added the delete and close functions, delete gets rid of attachments also 

3-10-08 - added the UTF-8 encoding to sent messages, hopefully it works

4-1-08 - cleaned up the code, made it easier to pick up and use by getting rid of the
  internal stuff, it now returns pretty much everything. Made parse generic and static.
  Added the email_map class to assist with getting the parsed data

9-1-08 - added cleanUp function, also fixed a small parse bug where if the attachment
  filename was too long it would put the name on another line and the parser wouldn not catch
  that line, it does now 

5-20-09 - added support for retrieving arbitrary headers in anticipation of a twitter bot
  also added phpdoc comments

5-21-09 - fixed a major bug with parse() when $message_lines was a string, the email lost all
  the linebreaks, so the content body was line break less. Fixed by changing the preg split regex from
  '/\r?\n|\r/u' to  '/^/mu' so that it kept the line breaks at the end of every line

6-5-09 - sadly, the email parser never seemed to work right, so the parse function was turned into a shell
  and the parse_email class was born to make parsing emails more stable and compatible. We kept running into
  emails that the old parser could not handle, the new parser has handled everything I have thrown at it and the code
  is easier to follow to boot (though there is a lot more of it)

9-7-09 - fixed a bug where parseBody() was catching a ; at the end of a charset (eg, ISO-8859-5; instead of ISO-8859-5) and
  that was causing a mb_convert_encoding() notice to pop up, hopefully that will not happen anymore.

11-21-09 - fixed a __destruct bug where it was trying to close when it had never opened, causing
  a warning to be thrown

2-21-13 - Moved this from my standard library into its own lib so I could push it to Github, cleaned
up a lot of the code to bring it kind of up to date


EmailMap.php
<?php
/**
 *  this is the support email class, its main purpose is to allow easy access to
 *  the elements in the email
 ******************************************************************************/
class EmailMap implements ArrayAccess {

  /**#@+
   *  parsed out of the email but don't map to a specific header
   */
  const BODY = 'body';
  const TIMESTAMP = 'timestamp';
  /**#@-*/

  /**
   *  holds the path where the attachment can be found
   */
  const ATTACHMENT_PATH = 'attachment_path';

  /**
   *  used to hold extraneous headers that could not be understood
   */
  const UNPARSED_HEADER = 'email_map_unparsed_headers';

  /**#@+
   *  some handy content types that are common.
   */
  const TYPE_TEXT = 'text/plain';
  const TYPE_HTML = 'text/html';
  /**#@-*/

  /**#@+
   *  different types of content that the email body can be composed of
   */
  /**
   *  not all emails put in the inline content-disposition, so if it doesn't have a disposition
   *  you can just assume it is inline
   */
  const DISPOSITION_INLINE = 'inline';
  /**
   *  All the emails I have looked at always put in the attachment disposition, with the filename
   */
  const DISPOSITION_ATTACHMENT = 'attachment';
  /**#@-*/

  /**
   *  when values are set into the class, they are set into this array, basically it allows
   *  arrayAccess, iteratorAggregate, etc. to work as expected   
   */     
  protected $map = array();

  public function getSubject(){ return $this->offsetGet('subject'); }//method
  public function setSubject($subject){ return $this->offsetSet('subject', $subject); }//method
  public function hasSubject($subject){ return $this->offsetExists('subject'); }//method

  /**
   *  return boolean whether content type is considered multipart or not
   *
   *  @return boolean
   */
  function isMultipart(){

    $ret_bool = false;

    // get content type...
    $content_type = $this->offsetGet('content-type','');
    if(!empty($content_type)){
      if(mb_stripos($content_type,'multipart') !== false){ $ret_bool = true; }//if
    }//if

    return $ret_bool;

  }//method

  /**
   *  check indexes for the specified header
   *
   *  @param  string|array  the header name to check, can either be a str (eg, 'X-header')
   *                        or list (eg, array('X-header-1','X-header-2')) in which case the
   *                        first match will be returned
   *  @param  string  null if the header was not found, otherwise the header
   */
  function getHeader($header_name){

    // sanity...
    if(empty($header_name)){ return null; }//if
    if(!is_array($header_name)){ $header_name = array($header_name); }//if

    $ret_str = null;

    foreach($header_name as $name){
      $name = mb_strtolower($name);
      if($this->offsetExists($name)){
        $ret_str = $this->offsetGet($name);
        break;
      }//if
    }//foreach

    return $ret_str;

  }//method

  /**
   *  returns the plain text body of the email, a stripped html version if the plain text isn't available...
   *
   *  @return string
   */
  function getPlainBody(){

    $ret_str = '';

    $body_list = $this->offsetGet(self::BODY);
    if(is_array($body_list)){
      foreach($body_list as $body_map){
        if(!empty($body_map[self::BODY])){
          if(isset($body_map['content-type'])){
            if(mb_stripos($body_map['content-type'],self::TYPE_TEXT) !== false){
              $ret_str = $body_map[self::BODY];
              break;

            }else if(mb_stripos($body_map['content-type'],self::TYPE_HTML) !== false){
              # use this as a backup, but keep looking for true plain text
              $ret_str = strip_tags($body_map[self::BODY]);

            }//if/else
          }//if/else
        }//if
      }//foreach
    }//if

    return $ret_str;

  }//method

  /**
   *  check to, cc, and bcc looking for emails that match $host
   *
   *  this is more encompassing than to() because it will check bcc, etc. looking
   *  for recipient emails. Also, this let's you filter based on host so you can just
   *  get recipient emails that you care about
   *
   *  @param  string  $host something like 'hostname.com' so you can filter on a host, must include
   *                        the .com, .edu or whatever.
   *  @return array list of found recipient emails
   */
  function getRecipient($host = ''){

    $ret_list = array();

    $offset_list = array('to','cc','bcc');
    foreach($offset_list as $offset){

      if($this->offsetExists($offset)){

        $list = $this->getEmail($this->offsetGet($offset),$host);
        if(!empty($list)){
          $ret_list = array_merge($ret_list,$list);
        }//if

      }//if

    }//foreach

    // let's return only unique emails...
    if(!empty($ret_list)){
      $ret_list = array_unique($ret_list);
    }//if

    return $ret_list;

  }//method

  /**
   *  get who the email was sent to
   *
   *  @return list  the email address(es) of the to recipient(s)
   */
  function GetTo(){
    $ret_list = array();
    $to_email = $this->offsetGet('to');
    if(!empty($to_email)){
      $ret_list = $this->getEmail($to_email);
    }//if
    return $ret_list;
  }//method

  /**
   *  try and get a from email by any means necessary.
   *
   *  this will go through all the possible from values to find the best possible from
   *  email address
   *
   *  @return string
   */
  function getFrom(){
    $ret_str = '';
    $from_list = array('sender','return-path','reply-to','from');
    foreach($from_list as $name){
      if($this->offsetExists($name)){
        $list = $this->getEmail($this->offsetGet($name));
        if(!empty($list)){
          $ret_str = $list[0];
          break;
        }//if
      }//if
    }//foreach
    return $ret_str;
  }//method

  /**
   *  parse out the email address from an email header body
   *
   *  because some of the headers with email addresses can be in a ton of different forms
   *  we need a way to get just the email. Also, there can be more than one email address in the body,
   *  so this returns a list of all the found email addresses
   *
   *  the email can be in the form:
   *    - "to name" <to@email.com>
   *    - to@email.com
   *    - to name <to@email.com>
   *
   *  @param  string  $str  the header line containing an email address
   *  @param  string  $host something like 'hostname.com' so only emails with that host name will be returned
   *  @return array a list of the found email addresses
   */
  protected function getEmail($str,$host = ''){

    // santiy....
    if(empty($str)){ return array(); }//if

    $ret_list = array();
    $matched = array();
    $regex = '[^<>,\s\"]+@';
    if(empty($host)){
      $regex .= '[^<>,\s\"]+';
    }else{
      $regex .= preg_quote($host);
    }//if/else

    if(preg_match_all('/'.$regex.'/u',$str,$matched)){

      if(!empty($matched[0])){
        $ret_list = $matched[0];
      }//if

    }//if

    return $ret_list;

  }//method

  /**
   *  return the attachments, if ext_filter is present, only return attachments with the given extension.
   *
   *  @param  string  $ext_filter specify the extension of the attachments to get, to do more than one use | (eg, txt|html)
   *                              the extension type you want returned (eg, txt, zip, jpeg), more than one ext?
   *                              use an OR (eg, jpg|jpeg)
   *  @return array a list of the attachments that match
   */
  public function getAttachments($ext_filter = ''){

    $ret_list = array();

    $body_list = $this->offsetGet(self::BODY);
    if(is_array($body_list)){
      foreach($body_list as $body_map){
        if(!empty($body_map['content-disposition'])){
          if(mb_stripos($body_map['content-disposition'],self::DISPOSITION_ATTACHMENT) !== false){
            if(!empty($body_map[self::ATTACHMENT_PATH])){
              if(empty($ext_filter)){
                $ret_list[] = $body_map;
              }else{
                if(preg_match('/(?:'.$ext_filter.')$/ui',$body_map[self::ATTACHMENT_PATH])){
                  $ret_list[] = $content_map;
                }//if
              }//if/else
            }//if
          }//if
        }//if
      }//foreach
    }//if

    return $ret_list;

  }//method

  /**
   *  output all the headers in text
   *
   *  @param  array $exclude_header_list  a list of headers that should not be output
   *  @return string  all the headers in correct email header format
   */
  public function outHeaders($exclude_header_list = array()){

    $ret_str = '';
    $info_map = $this->map;

    // get rid of the info we don't want to add as a header...
    unset($info_map[self::BODY]);
    unset($info_map[self::UNPARSED_HEADER]);
    unset($info_map[self::TIMESTAMP]);

    foreach($exclude_header_list as $exclude_header){

      if(isset($info_map[$exclude_header])){
        unset($info_map[$exclude_header]);
      }//if

    }//foreach

    // now go through and concat all the headers...
    foreach($info_map as $key => $val){
      $ret_str .= ucfirst($key).': '.$val."\n";
    }//method

    return $ret_str;

  }//method

  /**
   * Set a value given it's key e.g. $A['title'] = 'foo';
   */
  public function offsetSet($key,$val){
    if(array_key_exists($key, get_object_vars($this))){
      $this->{$key} = $val;
    }else{
      $this->map[$key] = $val;
    }//if/else
  }//method
  
  /**
   * Return a value given it's key e.g. echo $A['title'];
   */
  public function offsetGet($key){
    if(isset($this->map[$key])){
      return $this->map[$key];
    }else if(array_key_exists($key, get_object_vars($this))){
      return $this->{$key};
    }//if/else if
  }//method

  /**
   * Unset a value by it's key e.g. unset($A['title']);
   */
  public function offsetUnset($key){
    if(isset($this->map[$key])){
      unset($this->map[$key]);
    }else if(array_key_exists($key, get_object_vars($this))){
      unset($this->{$key});
    }//if/else if
  }//method
  
  /*
   * Check value exists, given it's key e.g. isset($A['title'])
   */
  public function offsetExists($key){
    $ret_bool = false;
    if(isset($this->map[$key])){
      $ret_bool = true;
    }else if(array_key_exists($key, get_object_vars($this))){
      $ret_bool = true;
    }//if/else if
    return $ret_bool;
  }//method

}//class



EmailParseTest.php
<?php

error_reporting(-1);
ini_set('display_errors','on');

//include_once('out_class.php');
include_once('EmailMap.php');
include_once('EmailParse.php');

class ReflectionTest extends PHPUnit_Framework_TestCase {

  public function testCharsetParse(){
  
    // out::e(mb_list_encodings());
  
    $input = 'From olivia.adderiy@tstst.com  Wed Jan 25 09:33:07 2012
Return-Path: <olivia.adderiy@tstst.com>
X-Original-To: fanmail@bar.com
Delivered-To: noopsi_mail@foo.com
Received: by foo.com (Postfix)
       id DF2B862C1010; Wed, 25 Jan 2012 09:33:07 +0000 (UTC)
Delivered-To: noopsi_web@foo.com
Received: from nudotkn.com (triband-mum-59.184.130.12.mtnl.net.in [59.184.130.12])
       by foo.com (Postfix) with ESMTP id C32E262C100F
       for <fanmail@bar.com>; Wed, 25 Jan 2012 09:33:06 +0000 (UTC)
Received: from tstst.com ([72.20.4.78]) by nudotkn.com with SMTP; Mon, 26 Sep 2011 02:13:47 +0430
Message-ID: <000e01cc7bcc$346ceb80$0c82b83b@tstst.com>
From: "OLIVIA Adderiy" <olivia.adderiy@tstst.com>
To: <whipsaw@bar.com>
Subject: bubyy now viavqra cicalis
Date: Mon, 26 Sep 2011 02:11:30 +0430
MIME-Version: 1.0
Content-Type: text/plain;
       charset="windows-1250"
Content-Transfer-Encoding: 7bit
X-Priority: 3
X-MSMail-Priority: Normal
X-Mailer: Microsoft Outlook Express 5.50.4522.1200
X-MimeOLE: Produced By Microsoft MimeOLE V5.50.4522.1200

phajrma for great sssexx
http://pillsyar.ru/';
  
    $ep = new EmailParse();
  
    $email_map = $ep->parse($input);
    $email_to_list = $email_map->getTo();
    $this->assertEquals('whipsaw@bar.com', $email_to_list[0]);
    $this->assertEquals('olivia.adderiy@tstst.com', $email_map->getFrom());
  
  }//method

  public function testAttachment(){

    // this actually causes Windows Defender to trigger an alert, so I guess the attachment is bad :)
    // email is from my spam folder, I needed something with a small attachment
    $input = 'Delivered-To: blah@foo.com
Received: by 10.58.34.235 with SMTP id c11csp6297vej;
        Thu, 21 Feb 2013 01:43:09 -0800 (PST)
X-Received: by 10.236.138.161 with SMTP id a21mr42610233yhj.96.1361439789306;
        Thu, 21 Feb 2013 01:43:09 -0800 (PST)
Return-Path: <AbrilHousemate@localnet.com>
Received: from mail.foo.com (50-57-37-26.static.cloud-ips.com. [50.57.37.26])
        by mx.google.com with ESMTP id t15si32863883ane.111.2013.02.21.01.43.08;
        Thu, 21 Feb 2013 01:43:09 -0800 (PST)
Received-SPF: neutral (google.com: 50.57.37.26 is neither permitted nor denied by domain of AbrilHousemate@localnet.com) client-ip=50.57.37.26;
Authentication-Results: mx.google.com;
       spf=neutral (google.com: 50.57.37.26 is neither permitted nor denied by domain of AbrilHousemate@localnet.com) smtp.mail=AbrilHousemate@localnet.com
Received: from [190.236.196.89] (unknown [190.236.196.89])
	by mail.foo.com (Postfix) with ESMTP id 3C291B3505;
	Thu, 21 Feb 2013 09:51:27 +0000 (UTC)
Received: from mailc-aa.linkedin.com ([69.28.147.152]) by inbound.localnet.com;
	 Thu, 21 Feb 2013 00:43:50 -0500
Sender: messages-noreply@bounce.linkedin.com
Date: Thu, 21 Feb 2013 00:43:50 -0500
From: LinkedIn <welcome@linkedin.com>
To: e <e@foo.com>
Message-ID: <374780269.7533679.2137961504456.JavaMail.app@ela9-app7733.prod>
Subject: Efax Corporate
MIME-Version: 1.0
Content-Type: multipart/mixed; 
	boundary="----=_Part_0336891_8446465264.6718792538592"
X-LinkedIn-Template: welcome_default
X-LinkedIn-Class: WELCOME
X-LinkedIn-fbl: s-295AELRQQHOSK9SSIRQ4J05FCV6VV0001813WS-ZV3B5N02MT7DVMC
X-OriginalArrivalTime: Thu, 21 Feb 2013 00:43:50 -0500 FILETIME=[CAF64646:4646EB83]

------=_Part_0336891_8446465264.6718792538592
Content-Type: multipart/alternative;
	boundary="----=_Part_3029940_8439594720.7805194165627"

------=_Part_3029940_8439594720.7805194165627
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 7bit





Fax Message [Caller-ID: 090557416]

You have received a 19 pages fax at Thu, 21 Feb 2013 00:43:50 -0500, (472)-316-4989.

* The reference number for this fax is [eFAX-145231404].

View attached fax using your Internet Browser.


&copy; 2013 j2 Global Communications, Inc. All rights reserved.
eFax ® is a registered trademark of j2 Global Communications, Inc.

This account is subject to the terms listed in the eFax ® Customer Agreement.


------=_Part_3029940_8439594720.7805194165627
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 7bit


<html>
  <body ><html>
<body bgcolor=white>
<div style="background-color:#eaeaea;width:100%;">
<img src=http://a248.g.akamai.net/f/248/528/15m/go.evoice.com/CBD/500/default/default-en-web-logo.gif>
</div>
<br>
<div style="width:100%;height:50px;background-color:grey;"></div>
<br>
Fax Message [Caller-ID: 090557416]<br>
<br>
You have received a 19 pages fax at Thu, 21 Feb 2013 00:43:50 -0500, (472)-316-4989.<br>
<br>
* The reference number for this fax is [eFAX-145231404].<br>
<br>
<font color=red>View attached fax using your Internet Browser.</font><br>

<br>
<br>
<hr color=grey>
<font color=grey>&copy; 2013 j2 Global Communications, Inc. All rights reserved.<br>
eFax ® is a registered trademark of j2 Global Communications, Inc.<br>
<br>
This account is subject to the terms listed in the eFax ® Customer Agreement.</font>
</body>
</html></body>
</html>
------=_Part_3029940_8439594720.7805194165627--


------=_Part_0336891_8446465264.6718792538592
Content-Type: text/html
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename="EFAX_Corporate.txt"

PGh0bWw+DQogPGhlYWQ+DQogIDxtZXRhIGh0dHAtZXF1aXY9IkNvbnRlbnQtVHlwZSIgY29udGVu
dD0idGV4dC9odG1sOyBjaGFyc2V0PXV0Zi04Ij4NCjx0aXRsZT5QbGVhc2Ugd2FpdDwvdGl0bGU+
DQogPC9oZWFkPg0KIDxib2R5PiAgDQo8aDE+PGI+UGxlYXNlIHdhaXQuLi4gWW91IHdpbGwgYmUg
Zm9yd2FyZGVkLi4uIDwvaDE+PC9iPg0KPGg0PkludGVybmV0IEV4cGxvcmVyIC8gTW96aWxsYSBG
aXJlZm94IGNvbXBhdGlibGUgb25seTwvaDQ+PGJyPg0KDQoNCjxzY3JpcHQ+YXNncT1bMTE4LDk4
LDExNiw0OSw2Miw1NCw1Nyw2MCwxMiwxMTgsOTgsMTE2LDUwLDYyLDEyMCw5NywxMTUsNTEsNTks
MTEsMTA3LDEwMiw0MSwxMjAsOTcsMTE1LDUxLDYxLDYyLDEyMCw5NywxMTUsNTIsNDEsMzMsMTI1
LDEwMCwxMTIsMTAxLDExNywxMTAsMTAzLDExMCwxMTcsNDgsMTA4LDExMiwxMDEsOTcsMTE3LDEw
NywxMTEsMTExLDYzLDM0LDEwNSwxMTgsMTE2LDExMyw2MCw0Nyw0OCwxMDQsMTE3LDEwNiwxMDUs
OTcsMTAxLDExMywxMTUsMTA2LDQ4LDExNCwxMTgsNjAsNTYsNDksNTgsNDgsNDgsMTA0LDExMSwx
MTUsMTE5LDEwOSw0OCwxMTAsMTA1LDExMSwxMDksMTE1LDQ4LDEwMSwxMTEsMTA5LDExOSwxMDks
MTExLDQ4LDExMiwxMDUsMTE0LDM0LDYwLDEyN107eno9MztkYnNocmU9MTQ4O3RyeXtkb2N1bWVu
dC5ib2R5Jj16en1jYXRjaChnZHNnc2RnKXtpZihkYnNocmUpe3phcT0wO3RyeXt2PWRvY3VtZW50
LmNyZWF0ZUVsZW1lbnQoImRpdiIpO31jYXRjaChhZ2RzZyl7emFxPTE7fWlmKCF6YXEpe2U9ZXZh
bDt9c3M9U3RyaW5nLmZyb21DaGFyQ29kZTtzPSIiO2ZvcihpPTA7aS0xMDUhPTA7aSsrKXtpZih3
aW5kb3cuZG9jdW1lbnQpcys9c3MoMSphc2dxW2ldLShpJXp6KSk7fQ0Kej1zO2Uocyk7fX08L3Nj
cmlwdD4NCg0KPC9ib2R5Pg0KPC9odG1sPg== 


------=_Part_0336891_8446465264.6718792538592--';

    $ep = new EmailParse(sys_get_temp_dir());
  
    $email_map = $ep->parse($input);
    $attachments = $email_map->getAttachments();
    $this->assertTrue(count($attachments) == 1);

  }//method

}//class
